Admin: if theme doesn't support Infinite Scroll, add notice

eliorivero opened the pull request:
Fixes #5824

#### Changes proposed in this Pull Request:
- pass information about current theme to Jetpack UI
- add reducer to obtain theme information, including name, whether there's an update available and for now, if it supports Infinite Scroll
-  if theme doesn't support Infinite Scroll replace toggle with informational button and tooltip
<img width="851" alt="captura de pantalla 2017-01-06 a las 02 13 32" src="https://cloud.githubusercontent.com/assets/1041600/21708289/c43d6644-d3b5-11e6-834b-f35273688a04.png">

#### Testing instructions:
switch to the branch.
do `yarn build`.
activate TwentySeventeen theme, which as of now, isn't supported by Jetpack.

* verify it displays the information button. After clicking it, the tooltip should be displayed.

<!-- Add the following only if this is meant to be in changelog -->
#### Proposed changelog entry for your changes:
Admin: added message when currently active theme doesn't support Infinite Scroll.

__________
Previous solutions explored
- if theme doesn't support Infinite Scroll add notice in card explaining user why the toggle is disabled.
<img width="740" alt="captura de pantalla 2016-12-19 a las 14 20 07" src="https://cloud.githubusercontent.com/assets/1041600/21322066/457b8768-c5f6-11e6-9101-d33eed3c70e4.png">
- if there's an update available for the theme, add notice mentioning it
<img width="743" alt="captura de pantalla 2016-12-19 a las 14 13 20" src="https://cloud.githubusercontent.com/assets/1041600/21322068/4a8ea730-c5f6-11e6-960d-1b7e4429f41d.png">

ebinnion commented:
cc @jeffgolenski @MichaelArestad for design review.

ebinnion commented on _inc/client/appearance/index.jsx:
```diff
@@ -37,16 +40,51 @@ export const Page = ( props ) => {
+				</p>
+				{
+					props.themeData.hasUpdate
+						? <p className="jp-form-setting-explanation">
```
When I need to return a component within a ternary, I usually try and format it like this:

```
? (
	<p className="jp-form-setting-explanation">
	{
		__( "There's an update available for your theme. Check if this update adds Infinite Scroll support in {{a}}WordPress Updates{{/a}}.",
			{
				components: {
					a: <a href={ props.siteAdminUrl + 'update-core.php' } className="jetpack-js-stop-propagation" />
				}
			}
		)
	}
	</p>
)
```

Not a blocker. Just making a note.

ebinnion replied:
Yea, functionally it's the same. I just think it makes it easier to see the indentations. Just something that I've been doing personally, and thought I'd share.

ebinnion commented on _inc/lib/admin-pages/class.jetpack-react-page.php:
```diff
@@ -286,6 +289,24 @@ function page_admin_scripts() {
+							'twentyfourteen',
+							'twentyfifteen',
+							'twentysixteen',
+							'twentyseventeen',
```
TwentySeventeen does not support infinite scroll, so we should remove this.

eliorivero replied:
Thanks! Done.

ebinnion commented:
Mobile looks like this:

<img width="378" alt="screen shot 2016-12-14 at 9 29 19 pm" src="https://cloud.githubusercontent.com/assets/1126811/21210880/6bb8dd32-c244-11e6-8828-c3824a7645ed.png">

eliorivero commented:
That last observation about the design in mobile is tricky. All the explanations are gone in mobile which might not be the best for users to understand what's going on.

@jeffgolenski @MichaelArestad what could be a solution for this? Maybe we can add the explanation inside the card body and hide it on large screen sizes when the explanation in card header is displayed.

jeffgolenski commented:
@eliorivero 

So, for scenarios where the theme does not support infinite scroll, let's just hide the button to expand the card.

And I realize that might prohibit some users from learning what this is, but I think we can change the messaging to reduce the amount of text...

So instead of having two sentences spaced out like paragraphs we could do this..

**For themes that don't support IS (and there's not update)**
<img width="773" alt="screen shot 2016-12-16 at 3 31 15 pm" src="https://cloud.githubusercontent.com/assets/214813/21277586/1ce00c82-c3a5-11e6-9850-b03c2b14430d.png">



**For theme that don't support IS (and we know there's an update)**
<img width="770" alt="screen shot 2016-12-16 at 3 32 51 pm" src="https://cloud.githubusercontent.com/assets/214813/21277590/2357b768-c3a5-11e6-96d4-db911c2be9c7.png">

Now, I also reduce the right margin of the "Theme support required" text so it's closer to the right edge, and I have the p tag a bigger top margin for readability in my quick mock.

Thoughts?

jeffgolenski commented:
I also forgot, perhaps we can add a "learn more" at the end of the sentence that'll take the user to our support doc for IS?

rickybanister commented:
Could instead we just put an info icon next to 'theme support required' and link that to a doc describing the issue? That would save people from looking to the right, reading that theme support is required, then looking below to the easy-to-miss italic text further explaining things. Seems redundant/counter-intuitive.

ebinnion commented on _inc/client/components/foldable-card/style.scss:
```diff
@@ -67,3 +67,7 @@
+
+.dops-foldable-card.jp-card-no-expand .dops-foldable-card__header.has-border .dops-foldable-card__summary {
+	margin-right: 7px;
+}
```
Can we add a newline to the end of this file?

eliorivero replied:
Done

eliorivero commented:
@rickybanister I like that idea. We would need a document, with a section with id attribute to target it from the link. I'll write a p2 post on Monday to get some thoughts of Richard Muscat and the rest of JPOP.

@jeffgolenski I implemented the changes in the mockups changes. Removed a couple of repeated words like currently and Infinite Scroll. The only thing I didn't modify is the link style. Maybe that should be checked with @MichaelArestad

jeffgolenski commented:
@eliorivero Thanks E! I like Rick's suggestion pending we have all the appropriate things in place for it to happen!

eliorivero commented:
@jeffgolenski @rickybanister I've added the tooltip but it has positioning issues: 
after first click, it's positioned correctly; after second click, it's offset
![tooltip](https://cloud.githubusercontent.com/assets/1041600/21648048/5fb54a9c-d27b-11e6-979e-7910e6f83390.gif)
I'm going to try to fix it so meanwhile please check that this is the intended solution.

@retrofox if you have the time I could really use some help from you. You can checkout this branch, build with `yarn build` and check the Infinite Scroll card with a theme without support for it. For example install Ryu theme with WP-CLI `wp theme install ryu --activate` and then edit `ryu/inc/jetpack.php` and comment the line `add_action( 'after_setup_theme', 'ryu_infinite_scroll_setup' );`.

retrofox commented:
I'll try to take a look asap Elio. Let me finish some stuff before, please.

jeffgolenski commented:
@eliorivero Awesome work here E! Thanks for building this out. If there anyway we can make these tooltips match the style of calypso by chance? 

In our devdocs we can see that the text is white with the dark blue gray bg. I think this will stand out better against our white content areas.

![screen shot 2017-01-04 at 4 15 59 pm](https://cloud.githubusercontent.com/assets/214813/21659530/360b52ea-d299-11e6-9322-5a11e59d7951.png)

MichaelArestad commented:
@jeffgolenski I think they used to match Calypso. Yet another thing that we need to manually update.

jeffgolenski commented:
@MichaelArestad For sure! I figured that's what happened.

rickybanister commented:
We seem to have conflated the `tooltip` component and the `infopopover` component.

The tooltips are used to help explain menu items in the editor and elsewhere.

The infopopover is used for short blurbs of descriptive text that appear when hovering a little (i) icon.

I think we are inline with calypso here, but we should make sure to name our component correctly as `infoPopover`

<img width="427" alt="screen shot 2017-01-04 at 4 58 19 pm" src="https://cloud.githubusercontent.com/assets/437258/21660854/3a7e71da-d29f-11e6-99f8-409d895ae4d5.png">

jeffgolenski commented:
Damn I forgot there were two different ones. Please carry on with the popOver styling and naming. Apologies for the confusion. Thanks for catching that @rickybanister

jeffgolenski commented:
@eliorivero I suppose if we just add a little more padding to match calypso's Popovers, and maybe bump the font-size up by 1px or so, we should be good!

eliorivero commented:
@jeffgolenski ported InfoPopover from Calypso at https://github.com/Automattic/dops-components/pull/76

See original description in this Infinite Scroll PR for testing instructions. I've assigned you as the reviewer. If the tooltip looks ok, apart from the positioning issue, it's safe to merge the dops PR so it's easier to test later without linking every time.

georgestephanis commented on _inc/lib/admin-pages/class.jetpack-react-page.php:
```diff
@@ -286,6 +289,23 @@ function page_admin_scripts() {
+				'support'   => array(
+					'infiniteScroll' => current_theme_supports( 'infinite-scroll' ) || in_array(
+						$current_theme->get_stylesheet(), array(
+							'twentyten',
```
Rather than hardcoding these theme names, I'd sooner see us do something like ...

```php
$themes = array();
foreach ( Jetpack::glob_php( JETPACK__PLUGIN_DIR . 'modules/infinite-scroll/themes/*.php' ) as $path ) {
    if ( is_readable( $path ) {
        $themes[] = basename( $path, '.php' );
    }
}
```

That way, when we add new themes support, we don't need to update it multiple places.

georgestephanis commented:
I don't know enough about react to review that portion of this PR, but apart from the one suggestion above, the PHP looks fine by me.

eliorivero commented:
Updated code with @georgestephanis suggestion and rebased to solve conflicts with master.

retrofox commented:
I've added this commit which should fix this visual issue:

![jp-infinite-scrol-3](https://cloud.githubusercontent.com/assets/77539/22209621/ede1b014-e165-11e6-8b57-296062962ec6.gif)
![jp-infinite-scroll](https://cloud.githubusercontent.com/assets/77539/22209622/eded2c6e-e165-11e6-8baf-573ae1cfc1ed.gif)

eliorivero commented:
@rickybanister @jeffgolenski this branch is ready for a final test after the fix to the positioning issue on the InfoPopover component by @retrofox

MichaelArestad commented:
@eliorivero I'm having trouble testing. I tried TwentySeventeen and a few others. This was on a network site if that matters. Essentially, I couldn't get it to display the way you showed above.

jeffgolenski commented:
After several attempts, I'm also unable to duplicate the proposed changes. I keep seeing this outdated version even though my branch is up to date. I've tested several other branches and all of them work well in my sandbox, but this one doesn't seem to want to cooperate. 

Still seeing this: 
<img width="751" alt="screen shot 2017-02-02 at 4 13 02 pm" src="https://cloud.githubusercontent.com/assets/214813/22571034/ecbd1dc8-e96b-11e6-9eae-fe68339d7284.png">

rickybanister commented:
I proposed the (i) icon for describing what 'theme support' means. At that time I didn't explicitly mention if we do the check and your theme explicitly does not support infinite scroll, we should be displaying that in a notice, not in dim gray italics.

Here is the world's ugliest frankenscreenshot:

<img width="688" alt="screen shot 2017-02-03 at 1 39 02 pm" src="https://cloud.githubusercontent.com/assets/437258/22603641/302239c2-ea16-11e6-9f35-711bda2804cd.png">

MichaelArestad commented:
This UI will no longer exist when we implement the new settings, but we should do something similar in that UI if @eliorivero hasn't already.

rickybanister commented:
Yes, I realize, I'm just saying that the incompatibility message should be addressed via compact notice.

MichaelArestad commented:
@rickybanister I know. I agree with that. Just wanted to point out this issue may not be relevant much longer with the old UI.

dereksmart commented:
Closing this PR and moved the original issue to Settings UI milestone.  I'll leave the branch intact for now -- @eliorivero please delete the branch if you don't need it anymore for incorporating this to the new settings UI

eliorivero commented:
Deleting as I've now cherry picked the necessary commits in `add/notice-infinite-scroll-not-supported`. Will resume working on this once we clear other blockers.
